File: predictor.R
library(sparklyr)
library(dplyr)
library(jsonlite)

config <- spark_config()
config$spark.yarn.access.hadoopFileSystems <- Sys.getenv("STORAGE")
config$spark.driver.memory <- "4g"

sc <- spark_connect(master = "local", config=config)

reloaded_model <- ml_load(sc, paste(Sys.getenv("STORAGE"),"/datalake/data/sentiment/lm_model_r",sep=""))

#args <- fromJSON('{"sentence":"Im no dunce"}')

predict_sentiment <- function(args) {
  test_text_df <- as.data.frame(args$sentence)
  colnames(test_text_df) <- "spoken_words"

  sdf_copy_to(sc, test_text_df, name="test_text", overwrite = TRUE)
  test_text <- tbl(sc, "test_text")

  test_text_tokenized <- test_text %>%
    mutate(spoken_words = regexp_replace(spoken_words, "[_\"\'():;,.!?\\-]", " ")) %>%
    ft_tokenizer(input_col="spoken_words",output_col= "word_list") %>%
    ft_stop_words_remover(input_col = "word_list", output_col = "wo_stop_words")

  test_text_w2v <- ft_word2vec(test_text_tokenized, input_col = "wo_stop_words", output_col = "result", min_count = 0)

  result <- ml_predict(reloaded_model,test_text_w2v)

  list("predicted_sentiment_value" = as.data.frame(result)$prediction)
}

File: predictor.py
from pyspark.sql import SparkSession
from pyspark.ml import PipelineModel
from pyspark.ml.feature import Word2VecModel, Tokenizer, StopWordsRemover
from pyspark.sql.functions import regexp_replace
import os

spark = SparkSession.builder \
      .appName("Sentiment") \
      .master("local[*]") \
      .config("spark.driver.memory","4g")\
      .config("spark.hadoop.yarn.resourcemanager.principal",os.getenv("HADOOP_USER_NAME"))\
      .getOrCreate()

s3_bucket = os.getenv("STORAGE")      

tokenizer = Tokenizer(inputCol="spoken_words", outputCol="word_list")
remover = StopWordsRemover(inputCol="word_list", outputCol="wo_stop_words")
w2v_model_fitted = Word2VecModel.load(s3_bucket + "/datalake/data/sentiment/w2v_model_fitted")
lr_model = PipelineModel.load(s3_bucket + "/datalake/data/sentiment/lr_model")

#args = {"sentence":"I'm no dunce, I was born an oaf and I'll die an oaf"}

def predict(args):
  input_sentence = args["sentence"]#.split(",")
  sentence_df = spark.createDataFrame([(input_sentence,)],['spoken_words'])
  sentence_df = sentence_df.select(regexp_replace('spoken_words',r'[_\"\'():;,.!?\\-]', ' ').alias('spoken_words'))
  sentence_df = tokenizer.transform(sentence_df)
  sentence_df = remover.transform(sentence_df)
  sentence_df = w2v_model_fitted.transform(sentence_df)
  result = lr_model.transform(sentence_df).collect()[0].prediction
  return {"result" : result}
